src/ReturnTypeProvider/ModelReturnTypeProvider.php

```diff
@@ -0,0 +1,93 @@
+use PhpParser\Node\Expr\New_;
+use PhpParser\Node\Expr\StaticCall;
+use PhpParser\Node\Name;
+use Psalm\CodeLocation;
+use Psalm\Context;
+use Psalm\Internal\MethodIdentifier;
+use Psalm\Plugin\Hook\MethodExistenceProviderInterface;
+use Psalm\Plugin\Hook\MethodReturnTypeProviderInterface;
+use Psalm\StatementsSource;
+use Psalm\Type;
+use Psalm\Type\Atomic\TGenericObject;
+use Psalm\Type\Atomic\TNamedObject;
+use Psalm\Type\Atomic\TTemplateParam;
+use Psalm\Type\Union;
+
+final class ModelReturnTypeProvider implements MethodReturnTypeProviderInterface
+{
+
+    public static function getClassLikeNames(): array
+    {
+        return [Model::class];
+    }
+
+    public static function getMethodReturnType(StatementsSource $source, string $fq_classlike_name, string $method_name_lowercase, array $call_args, Context $context, CodeLocation $code_location, array $template_type_parameters = null, string $called_fq_classlike_name = null, string $called_method_name_lowercase = null)
+    {
+        if (!$source instanceof \Psalm\Internal\Analyzer\StatementsAnalyzer) {
+            return Type::getMixed();
+        }
+
+        if ($method_name_lowercase === '__call' || $method_name_lowercase === '__callStatic') {
+            // proxy to builder
+            //dump('proxying,.,.,');
+            // @todo: proxy the method call onto this typed eloquent builder
+            //dd("getMethodReturnType: fq_name: $fq_classlike_name called_name: $called_fq_classlike_name | $method_name_lowercase");
+            return self::proxyToBuilder(...func_get_args());
+            /**
+            return new Union([
+                new TGenericObject(Builder::class, [
+                    new Union([new TTemplateParam('TModel', new Union([new TNamedObject($called_fq_classlike_name)]), $called_fq_classlike_name)]),
+                ]),
+            ]);
+             * */
+        }
+    }
+
+    private static function proxyToBuilder(\Psalm\Internal\Analyzer\StatementsAnalyzer $statementsAnalyzer, string $fq_classlike_name, string $method_name_lowercase, array $call_args, Context $context, CodeLocation $code_location, array $template_type_parameters = null, string $called_fq_classlike_name = null, string $called_method_name_lowercase = null)
+    {
+        $oldNodeData = $statementsAnalyzer->node_data;
+        $statementsAnalyzer->node_data = clone $statementsAnalyzer->node_data;
+        $statementsAnalyzer->getSource()->getTemplateTypeMap()
+        $suppressed_issues = $statementsAnalyzer->getSuppressedIssues();
+
+        if (!in_array('PossiblyInvalidMethodCall', $suppressed_issues, true)) {
+            $statementsAnalyzer->addSuppressedIssues(['PossiblyInvalidMethodCall']);
+        }
+
+        $was_inside_call = $context->inside_call;
+
+        $context->inside_call = true;
+
+        // call here
+        // todo: right now  this is just for static
+
+        $fakeMethodCall = new StaticCall(new Name(Builder::class), $method_name_lowercase, $call_args);
```

Comment:
@muglug I'm struggling with this. How do we fake a method call to a generic object? How do I pass the template parameters? Would love any tips you may have!

Essentially I want to say that  `__call` is a proxy to `Builder<User>`

Reply:
Meanwhile working on a stubbing approach...

Reply:
Mind opening a ticket in vimeo/psalm for that?